src/libopensc/card-piv.c

```diff
@@ -467,205 +791,1921 @@ static struct sc_card_driver piv_drv = {
- * Used by  GET DATA, PUT DATA, GENERAL AUTHENTICATE
- * and GENERATE ASYMMETRIC KEY PAIR.
- * GET DATA may call to get the first 128 bytes to get the length from the tag.
+ *  Construct SM protected APDU
  */
 
-static int piv_general_io(sc_card_t *card, int ins, int p1, int p2,
-	const u8 * sendbuf, size_t sendbuflen, u8 *recvbuf,
-	size_t recvbuflen)
+static int piv_encode_apdu(sc_card_t *card, sc_apdu_t *plain, sc_apdu_t *sm_apdu)
 {
-	int r;
-	sc_apdu_t apdu;
+	int r = 0;
+	piv_private_data_t * priv = PIV_DATA(card);
+	cipher_suite_t *cs = priv->cs;
+	u8 pad[16] ={0x80};
+	u8 zeros[16] = {0x00};
+	u8 IV[16];
+	u8 header[16];
+	int padlen = 16; /* may be less */
+	u8 *sbuf = NULL;
+	size_t sbuflen = 0;
+	int MCVlen = 16;
+	int enc_datalen = 0;
+	int T87len;
+	int T97len = 2 + 1;
+	int T8Elen = 2 + 8;
+
+	int outli = 0;
+	int outl = 0;
+	int outll = 0;
+	int outdl = 0;
+	u8 discard[16];
+	int macdatalen;
+	size_t C_MCVlen = 16; /* debugging*/
+
+	u8 *p;
+	EVP_CIPHER_CTX *ed_ctx = NULL;
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+	CMAC_CTX *cmac_ctx  = NULL;
+#else
+	EVP_MAC_CTX *cmac_ctx = NULL;
+	EVP_MAC *mac = NULL;
+	OSSL_PARAM cmac_params[2];
+	size_t cmac_params_n;
+#endif
 
 	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
 
-	r = sc_lock(card);
-	if (r != SC_SUCCESS)
-		LOG_FUNC_RETURN(card->ctx, r);
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+	cmac_ctx = CMAC_CTX_new();
+	if (cmac_ctx == NULL) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+#else
+	mac = EVP_MAC_fetch(NULL, "cmac", NULL);
+	cmac_params_n = 0;
+	cmac_params[cmac_params_n++] = OSSL_PARAM_construct_utf8_string("cipher", cs->cipher_cbc_name, 0);
+	cmac_params[cmac_params_n] = OSSL_PARAM_construct_end();
+	if (mac == NULL
+			|| (cmac_ctx = EVP_MAC_CTX_new(mac)) == NULL) {
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+#endif
 
-	sc_format_apdu(card, &apdu,
-			recvbuf ? SC_APDU_CASE_4_SHORT: SC_APDU_CASE_3_SHORT,
-			ins, p1, p2);
-	apdu.flags |= SC_APDU_FLAGS_CHAINING;
-	apdu.lc = sendbuflen;
-	apdu.datalen = sendbuflen;
-	apdu.data = sendbuf;
+	ed_ctx = EVP_CIPHER_CTX_new();
+	if (ed_ctx == NULL) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
 
-	if (recvbuf && recvbuflen) {
-		apdu.le = (recvbuflen > 256) ? 256 : recvbuflen;
-		apdu.resplen = recvbuflen;
+	if (EVP_EncryptInit_ex(ed_ctx, (*cs->cipher_ecb)(), NULL, priv->sm_session.SKenc, zeros) != 1
+			|| EVP_CIPHER_CTX_set_padding(ed_ctx,0) != 1
+			|| EVP_EncryptUpdate(ed_ctx, IV, &outli, priv->sm_session.enc_counter, 16) != 1
+			|| EVP_EncryptFinal_ex(ed_ctx, discard, &outdl) != 1
+			|| outdl != 0) {
+		sc_log(card->ctx,"SM encode failed in OpenSSL");
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+
+	sm_apdu->cla = 0x0c;
+	sm_apdu->ins = plain->ins;
+	sm_apdu->p1 = plain->p1;
+	sm_apdu->p2 = plain->p2;
+	/* all APDU become cse=4  cse=1 stay short. May support extended in future */
+	sm_apdu->cse = (plain->cse & ~0x07) + 4;
```

Comment:
This has a comment, but still the operation is very obfuscated. Can we use some constants, simple `if` condition or ternary operator to make it more readable what you do here?

Reply:
Fixed in d207726771accc65dfd1ad71f3ec5e40d9b04c0c Is this more acceptable?

Reply:
Yes. Thanks!